tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,1176 @@
+ * a correct ratio to each server in the group.
+ * 1. If the minimal ratio is 1, then fill the entries with minimal ratio
+ *    with values from an entry with the maximum ratio. Fill the entries
+ *    with maximum ratio with values from an entry with minimal ratio.
+ * 2. Sort the resulting array by ratio in descending order as required
+ *    by the scheduling algorithm.
+ * 3. Select the part of the array that omits entries from step 1 if any.
+ *    Those are entries at the start and at the end of the array. Reverse
+ *    the sequence of server descriptor indices in that part of the array.
+ *    The resulting pairing of servers to ratios is the target.
+ */
+static void
+__tfw_sched_ratio_calc_dynamic(TfwRatioPool *rpool, TfwRatio *ratio)
+{
+	size_t si, max_val_idx = 0, left = 0, right = 0;
+	unsigned int max_ratio = 0, has_one_val = 0;
+	TfwRatioSrvData *srvdata = ratio->srvdata;
+
+	/* Calculate ratios based on server RTT values. */
+	if (!tfw_sched_ratio_calc(rpool, ratio, &max_val_idx))
+		return;
+
+	/*
+	 * It's guaranteed here that NOT all calculated ratio values are
+	 * equal. See if there are ratio values that equal to 1. If so,
+	 * do actions described in step 1 in the function's description.
+	 * Adjust the sum of ratios that is changed in this procedure.
+	 */
+	for (si = 0; si < rpool->srv_n; ++si) {
+		if (srvdata[si].oratio == 1) {
+			has_one_val = 1;
+			break;
+		}
+	}
+	if (has_one_val) {
+		unsigned long orsum = ratio->schdata.orsum;
+		TfwRatioSrvData sdent_one = srvdata[si];
+		TfwRatioSrvData sdent_max = srvdata[max_val_idx];
+
+		/* Save maximum ratio value for future use. */
+		max_ratio = srvdata[max_val_idx].oratio;
+
+		for (si = 0; si < rpool->srv_n; ++si) {
+			if (srvdata[si].oratio == 1) {
+				srvdata[si].weight = sdent_max.weight;
+				srvdata[si].oratio =
+				srvdata[si].cratio = sdent_max.oratio;
+				orsum += sdent_max.oratio - 1;
+			} else if (srvdata[si].oratio == sdent_max.oratio) {
+				srvdata[si].weight = sdent_one.weight;
+				srvdata[si].oratio =
+				srvdata[si].cratio = sdent_one.oratio;
+				orsum -= sdent_max.oratio - 1;
+			}
+		}
+		ratio->schdata.crsum = ratio->schdata.orsum = orsum;
+	}
+
+	/* Sort server data entries by ratio in descending order. */
+	sort(srvdata, rpool->srv_n, sizeof(srvdata[0]),
+	     tfw_sched_ratio_srvdata_cmp, tfw_sched_ratio_srvdata_swap);
+
+	/*
+	 * Do actions described in step 3 in the function's description.
+	 * Select the part of the array that omits entries from step 1
+	 * if there are any. Those are entries at the start and at the
+	 * end of the array. Reverse the sequence of server descriptor
+	 * indices in that part of the array.
+	 */
+	if (!has_one_val) {
+		left = 0;
+		right = rpool->srv_n - 1;
+	} else {
+		for (si = 0; si < rpool->srv_n; ++si)
+			if (srvdata[si].oratio == max_ratio) {
+				left = si + 1;
+			} else if (srvdata[si].oratio == 1) {
+				right = si - 1;
```

Comment:
If we're here after a call to `tfw_sched_ratio_calc()`, then it's guaranteed that not all calculated ratios are equal. There's a comment on that on line 344 above. That's the precondition for going forward with these manipulations. So, if we're right here in this code block, then there are at least two servers, and one of them has ratio of 1, and there's another (different) server with maximum ratio.